### PR TITLE
Build with CMake

### PR DESCRIPTION
**Pull request by ridiculousfish**
**TLDR: let's build fish with CMake and eventually replace autotools, so we can drop a ton of cruft.** This PR has enough for day-to-day development with CMake. I'm looking forwards to feedback and discussion.

## What is CMake?

CMake is a meta-build system. At a high-level it plays the same role as ./configure by generating a Makefile, but it can also output other builds (Ninja, CLion, XCode, CodeBlocks VS, and more).

CMake is mature and battle-tested. It's used by major OS projects like LLVM, KDE, MySQL. It has great docs and is easy to find help on StackOverflow, etc.

KDE switched to CMake, [here's their story](https://dot.kde.org/2007/02/22/road-kde-4-cmake-new-build-system-kde).

Probably you already know CMake and hopefully don't need to be sold on it. If not there's info on CMake advantages at the bottom.

## What does this PR include?

This PR:

1. Builds fish, fish\_indent, fish\_key\_reader
2. Builds and runs all tests
3. Builds the docs and man pages
4. Can install fish and all the rest (web\_config, etc)

What's missing:

1. Generating release tarballs, .deb files, macOS installer packages, etc
2. Integration with TravisCI / OBS
3. New build instructions
4. The "helpful hints" that the Makefile prints at the end

## Try it!

Easy peasy:

1. `git checkout` this PR.
2. Make a build directory: `mkdir /tmp/fish-build ; cd /tmp/fish-build`
3. Generate a Makefile: `cmake /path/to/fish-shell`
4. `make test`
5. `mkdir /tmp/fish-install ; env DESTDIR=/tmp/fish-install make install`

`make VERBOSE=1` for a verbose build.

Advanced command line users will like [Ninja](https://ninja-build.org):

1. `cmake -G Ninja /path/to/fish-shell; and ninja`

Or if you want Xcode:

1. `cmake -G Xcode /path/to/fish-shell`

## What happens if we merge this?

Here's a sketch of how this could go:

1. The autotools build system will continue to work until the CMake build is at 100%. Devs can use CMake or the existing autotools system as they prefer. During this time, we'll have to duplicate build system changes across both autotools and CMake.
2. We'll teach the CMake build how to generate packages, [presumably with CPack](https://cmake.org/Wiki/CMake:Packaging_With_CPack).
3. We'll incrementally switch Travis, OBS, homebrew, etc. to using CMake.
4. We'll remove fish.xcodeproj and most of the stuff in osx/.
5. We'll remove configure.ac, Makefile.in, m4, and all the rest of the autotools stuff. 
6. (Longer term) we can remove pcre2 and instead rely on CMake's package management. Also clean up the docs build and muParser build to make them more CMake friendly.

## How does the CMake build handle:

1. **PCRE2**: This PR uses the system PCRE2 if one is installed. If not, it builds the bundled PCRE2 via CMake (which PCRE2 conveniently supports). Once we remove support for autotools we can stop bundling pcre2 and use the CMake package management.

2. **muParser**: This builds muParser as an [ExternalProject](https://cmake.org/cmake/help/v3.0/module/ExternalProject.html) using the muParser autotools build. The path forward here is to fork muParser, and add our own CMake build system.

3. **Docs**: The docs are by far the most complicated part of fish build. The autotools path uses a bunch of sed commands and relies on many temporary files. This PR factors those commands into separate shell scripts under build\_tools, which both the Makefile and CMake invoke.

    This is admittedly gross, but the docs build was already gross. It should be factored into a monolithic script that builds everything, instead of granular individual targets.

### Mature, battle-tested, widely available

This is the main advantage that CMake has over gyp, boost.build, scons, etc. No build system is as ubiquitous as autotools, but CMake is number two.

### Much simpler builds

autotools builds are famously difficult to maintain: arcane m4 syntax, workarounds for ancient systems, layers upon layers. CMake's syntax is a little weird, but simple and regular.

Nested projects are very common and easy with CMake. It is very simple to build PCRE2 as a nested project using PCRE2's CMake path. No more "recursive Make considered harmful."

CMake also knows about development artifacts such as header files and libraries. autotools must be manually taught header dependencies ("make depend"), filename extensions (.so vs .dylib), so on. 

### Actual package management

CMake has built in support for [external dependencies](https://cmake.org/cmake/help/latest/module/ExternalProject.html). It knows how to look for them locally, fetch them via git, etc. PCRE2 would have been much simpler to add to fish if we were using CMake.

### Replaces XCode project

Unlike Make, CMake knows about macOS-isms like [bundles](https://cmake.org/cmake/help/latest/prop_tgt/MACOSX_BUNDLE.html#prop_tgt:MACOSX_BUNDLE) and [deployment targets](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_DEPLOYMENT_TARGET.html). We can support macOS's needs without maintaining the separate XCode project and osx/ contents.

CMake can also generate CLion or VS projects, so we can support a lot more IDEs.

### Out-of-source builds

CMake builds in a different directory from the source code. This takes a little bit of getting used to [but is actually great](http://voices.canonical.com/jussi.pakkanen/2013/04/16/why-you-should-consider-using-separate-build-directories/). Your source directory stays clean. `make distclean` becomes `rm -Rf *`. Multiple builds (debug vs release) are trivial.

This is possible in princple with autotools, but takes a lot of work.

## Disadvantages of CMake

1. This will add cmake as a dependency for building fish. cmake is widely available but less likely to be installed compared to GNU Make.
2. CMake's generated Makefiles are not portable by design (e.g. have hard-coded paths), so our tarballs will require CMake to build. Tarball users won't be able to just `./configure ; make ; make install`, unless we provide a wrapper.
3. CMake is not quite as flexible as GNU Make. For example, it is hard to implement the `active_test_goals` hack from make, which serializes tests but only those that are to be run. However CMake has good support for invoking external scripts, so it ought to be flexible enough.

### CODE REVIEW

**Comment by mqudsi** [comment]
Oh my gosh, thank you! 

I've locally been building fish with meson for its ninja backend, but that has the extremely unfortunate side effect of introducing python as a build requirement (the meson team feels that's OK, [though I disagreed](https://github.com/mesonbuild/meson/issues/2335)).

cmake and meson would piping to ninja would both generate far smarter build files that do not need to rebuild as much or as often when jumping between revisions to test when/where something broke.

With regards to your point:

> CMake's generated Makefiles are not portable by design (e.g. have hard-coded paths), so our tarballs will require CMake to build. Tarball users won't be able to just ./configure ; make ; make install, unless we provide a wrapper.

I would argue that this is a dealbreaker, and we shouldn't release without end users being able to `./configure; make; sudo make install;` as they've been able to for years with pretty much any FOSS project. It's why I never even whispered about my meson fish builds (despite spending a day replacing the autoconf script with its meson equivalent, feature tests included) as changing that workflow was (for me) a nonstarter.

**Comment by ridiculousfish** [comment]
> I would argue that this is a dealbreaker, and we shouldn't release without end users being able to ./configure; make; sudo make install

I wonder how other OSS projects handle that. It looks like libpng and pcre2 both manually maintain CMake and autotools systems, which seems nuts. [bro](https://www.bro.org) has ./configure and Makefile wrappers that call out to CMake, which seems OK. It might be worth asking on StackOverflow or CMake mailing list.

**Comment by floam** [comment]
Super cool!

**Comment by mqudsi** [comment]
configure-cmake looks interesting, nice find.

I was actually thinking of a configure script that would take take a cmake-generated makefile and plug in `prefix` and co without needing to shell out to `cmake` at all (so it's only a dev and not a build dependency) via some sed voodoo, but I'm not sure how practical that is (presumably it would have been done by others if it were easy?).

neovim wraps cmake in a Makefile that shells out to cmake and ninja, no `./configure` required (cmake build time dependency though, of course): https://github.com/neovim/neovim/blob/master/Makefile

(When I was deciding between meson, cmake, and other "modern" build systems for $work I came across [tup](http://gittup.org/tup/) which is downright awesome (and generates dependency-free build scripts for distribution tarballs) and written by folk that obviously have their head on straight, but it has a dev (not build) dependency on fuse (which is how it operates; inspecting which files are opened during build and automatically configuring them as dependencies), but the fuse dependency is a huge non-starter and makes Windows support (even via WSL) out of the question.)

**Comment by faho** [comment]
>I've locally been building fish with meson for its ninja backend

Hehe... so we've had at least two people trying out meson.

I did as well, and [here's what I wrote up then](https://gitter.im/fish-shell/fish-shell?at=590b51174d51a79e643c2660):

>So, I've got fish building and installing with meson. It takes about half the time autotools takes, and I find the file much more readable than both configure.ac and Makefile.in. However, documentation is sparse and it won't build on travis (since even the "experimental" Ubuntu they use is too old - the regular one is actually been EOL'd last week).

>The remaining issues:

>- Localization seems to require a file that lists all files to be localized
>- Version handling is weird - they have a "vcs_tag" function to get a version via git, but that won't expose it as a variable, only writing it to a file
>- Incorporating doxygen likely requires scripting
>- This would exclude Debian Jessie (i.e. the current one IIRC) and Ubuntu < 16.04, plus possibly many others.

> So all-in-all, unless there's interest in a second build system for development (i.e. test with meson, keep autotools for distributions), I think I'm abandoning this experiment for now. If anyone cares, see my meson branch.

So unlike @mqudsi, I never really had an issue with the python dependency, though my concern is mostly linux systems, and those tend to have python anyway.

What I did have an issue with is that meson is still so new that it's adding functionality left and right, so it's kinda hard to use it in such a way that it still works on old OSen - if they have it at all.

------

So, as you might see, cmake in this implementation solves all of those issues - it already has the doc stuff (that I never really attempted since I gave up after figuring out that it was a non-starter for Debian), the version generation works, cmake has "GLOB" to avoid specifying every .po file and .fish script. It can also still use ninja, which manages to make the build quite a bit faster (though cmake's makefiles are apparently bad, since a cmake-with-make build takes 50% longer on my system than with our autoconf one).

So, while I like meson's syntax a bit more than cmake (and I dislike having to explicitly pick ninja), cmake solves some real problems we'd have with it.

It's also just simply more well-known than meson.

(Thwough of course it's possible that @mqudsi's meson work is better than mine, since I am not well-versed in build systems)

-----

So, I am apparently not alone in not loving autostuff/make. IMHO, the main advantage it has is that it's ubiquitous. However, cmake is also available on every linux distribution, and has been for quite a while. It's available on windows and I'd imagine the BSDs. I don't think availability is going to be an issue.

There is one thing that I dislike in pretty much anything that isn't autotools, and that's configuring directories. I'm currently rewriting my PKGBUILD (the thing that gives me nice archlinux packages to install) to use cmake if available, and I find it really hard to pass the right things. I.e. 

```
cmake -G Ninja -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_DATADIR=/share -DCMAKE_INSTALL_DOCDIR=/share/doc/fish ..
```

is noticeably uglier than and doesn't work as a replacement for 

```
./configure --prefix=/usr --sysconfdir=/etc --docdir=/usr/share/doc/fish
```

**Comment by floam** [comment]
For some reason, the binary generated with CMake/ninja is way larger than the autotools build, like 4595444 bytes vs. 2066896 bytes.

**Comment by floam** [comment]
Surprised that the CMake build is linking libform, while the autotools build isn't:

```
Load command 13
          cmd LC_LOAD_DYLIB
      cmdsize 56
         name /usr/lib/libform.5.4.dylib (offset 24)
   time stamp 2 Wed Dec 31 16:00:02 1969
      current version 5.4.0
compatibility version 5.4.0
Load command 14
```

**Comment by ridiculousfish** [comment]
@floam the CMake build by default is -O0, the autotools build is -O2. Can you repeat your experiment except run `cmake -DCMAKE_BUILD_TYPE=Release /path/to/fish-shell`?

`libform` is probably being picked up by CMake's FIND_CURSES implementation. Nice idea to check the libs!

**Comment by floam** [comment]
Yeah, it was just the optimization level. Release build is way smaller.

**Comment by mqudsi** [comment]
Do we want to set the default level to the same we currently use on CMake?

**Comment by floam** [comment]
How about things like `make style`, and `make iwyu`? Should cmake provide an interface or should we just execute shell scripts directly?

**Comment by mqudsi** [comment]
@zanchey what do you think about the switch to CMake? 

If we have consensus, we should probably merge these changes to a cmake branch upstream so we can make PRs against it.

**Comment by ridiculousfish** [comment]
@mqudsi IMO the CMake default of -O0 is better for day to day development: faster builds and easier debugging. I'd like to keep it.

If we decide to go with this I don't think we need an upstream branch, we can just merge to master. This doesn't break the autotools build.

**Comment by zanchey** [comment]
SGTM!

**Comment by mqudsi** [comment]
@ridiculousfish does the cmake build also depend on pkg-config? I keep forgetting to double check and then update the readme to note that building fish via autotools currently indirectly depends on pkg-config.

**Comment by floam** [comment]
We depend on pkg-config? I don't have it and have never had trouble with anything.

**Comment by mqudsi** [comment]
I recall being unable to build on solaris without installing it. But like I said, I'd have to double-check.

That said, I'm surprised you don't have it installed; it usually ends up being installed regardless of the package manager you use on most modern *nix systems.

**Comment by floam** [comment]
I use MacOS, FWIW.

**Comment by mqudsi** [comment]
Hmmm, reviewing my logs it seems that I did have to install pkg-config but removing it now allows fish to build. Perhaps I need it to build a dependency or something... no matter.

**Comment by floam** [comment]
Looking over FindCurses, it looks like `${CURSES_LIBRARIES}` contains all of `${CURSES_LIBRARY}`, `${CURSES_EXTRA_LIBRARY}` (libtinfo), and `${CURSES_FORM_LIBRARY}` that it can find. This seems to make it stop linking the unnecessary libform library here:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index e4628bc7..9501887b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -90,7 +90,7 @@ ENDFUNCTION(FISH_LINK_DEPS)
 ADD_LIBRARY(fishlib STATIC ${FISH_SRCS})
 TARGET_SOURCES(fishlib PRIVATE ${FISH_HEADERS})
 TARGET_LINK_LIBRARIES(fishlib
-  ${CURSES_LIBRARIES} Threads::Threads ${CMAKE_DL_LIBS}
+  ${CURSES_LIBRARY} ${CURSES_EXTRA_LIBRARY} Threads::Threads ${CMAKE_DL_LIBS}
   ${PCRE2_LIB} muparser)
 
 # builtin_math.cpp needs to see muParser's built header.
```

**Comment by ridiculousfish** [comment]
Thanks @floam, added that as 2ca7b88d4945ffae6579e1c0a5c6f8c5e0106db2

**Comment by zanchey** [comment]
One problem I've run into is that the CMakeFiles directory is used by CMake for the actual build process, so in-tree builds (i.e. `cmake .`) fail because the include files get wiped out by the build process.

`git mv CMakeFiles cmake_includes` and related fixups make in-tree builds work.

**Comment by ridiculousfish** [comment]
I didn't know in-tree builds were even a possibility outside of a dedicated ./build directory, but if it's easy to fix let's do it!

**Comment by zanchey** [comment]
I have a patch series in the works that fixes that, among other things.

**Comment by zanchey** [comment]
See #4535 for further work.

Still remaining:
 - [ ] checking for fixed-argument `tparm` (the old "Solaris kludge"), which is still required on NetBSD unless we add a `TPARM_VARARGS` define, and possibly on IllumOS or other Solaris-derived systems
 - [ ] checking for and enabling Gettext - at present `USE_GETTEXT` is always off
 - [ ] supporting `make install` without Doxygen

**Comment by zanchey** [comment]
AFAICT the only thing still requiring autotools is the build_tarball script
